### PR TITLE
Solved last master build system changes to EmitBinPath

### PR DESCRIPTION
**Pull request by lmarzocchetti**
From the last master changes of the commit: "dffc8c4", they changed "emit_bin_path" to "emit_digest", so the build of zls was broken for the latest master, updating these result in a succeful build

### CODE REVIEW

**Comment by lmarzocchetti** [comment]
i don't know why fails on macos or windows. I'm on linux, i've tested the build on macos and it works perfectly, but the failed test still remain

File: src/ZigCompileServer.zig [hunk 1]
```diff
@@ -59,9 +59,9 @@ pub fn receiveMessage(client: *Client) !InMessage.Header {
     return error.Timeout;
 }
 
-pub fn receiveEmitBinPath(client: *Client) !InMessage.EmitBinPath {
+pub fn receiveEmitBinPath(client: *Client) !InMessage.EmitDigest {
```
[the remark this answers is not in the document]
Comment (lmarzocchetti) [reply src/ZigCompileServer.zig:62]: I think yes, now i'm outside (live in italy), so when i come home i try to rename functions and where it calls them

**Comment by Techatrix** [comment]
The relevant zig commit is https://github.com/ziglang/zig/commit/43f73af3595c3174b8e67e9f2792c3774f2192e9.
The reason why `emit_bin_path` has been renamed to `emit_digest` is because the messages no longer sends the entire file path.
> This is fixed by communicating the final hash of the artifact path (the hash of the corresponding /o/<hash> directory) and communicating this instead of the entire path. This changes the zig build --listen protocol to communicate hashes instead of paths, and emit_bin_path is accordingly renamed to emit_digest.

---

> i don't know why fails on macos or windows. I'm on linux, i've tested the build on macos and it works perfectly, but the failed test still remain

Did you run `zig build test` or just `zig build`? `zig build test` should fail on linux as well. I believe the reason why CI passed on linux is because the random hash just happened to resemble a "valid" file path. re-running CI makes it [fail](https://github.com/zigtools/zls/actions/runs/10488644839/job/29061279867?pr=2007) as expected.

**Comment by lmarzocchetti** [comment]
Ok, i'll take a look at this commit and the test and see if i can solve this

**Comment by Techatrix** [comment]
> Ok, i'll take a look at this commit and the test and see if i can solve this

:+1: Feel free to let me know if there is anything I can do to help you with this.

File: src/ZigCompileServer.zig (hunk 1 above)
[the remark this answers is not in the document]
Comment (moderation) [reply src/ZigCompileServer.zig:62]: This diff working for me:
```diff
diff --git a/src/ZigCompileServer.zig b/src/ZigCompileServer.zig
index 1441683..d11f3ff 100644
--- a/src/ZigCompileServer.zig
+++ b/src/ZigCompileServer.zig
@@ -59,9 +59,9 @@ pub fn receiveMessage(client: *Client) !InMessage.Header {
     return error.Timeout;
 }

-pub fn receiveEmitBinPath(client: *Client) !InMessage.EmitBinPath {
+pub fn receiveEmitDigest(client: *Client) !InMessage.EmitDigest {
     const reader = client.pooler.fifo(.in).reader();
-    return reader.readStruct(InMessage.EmitBinPath);
+    return reader.readStruct(InMessage.EmitDigest);
 }

 pub fn receiveErrorBundle(client: *Client) !InMessage.ErrorBundle {
diff --git a/src/translate_c.zig b/src/translate_c.zig
index e83636e..b8f5ff9 100644
--- a/src/translate_c.zig
+++ b/src/translate_c.zig
@@ -223,11 +223,11 @@ pub fn translate(
                 // log.debug("zig-version: {s}", .{zcs.receive_fifo.readableSliceOfLen(header.bytes_len)});
                 zcs.pooler.fifo(.in).discard(header.bytes_len);
             },
-            .emit_bin_path => {
-                const body_size = @sizeOf(std.zig.Server.Message.EmitBinPath);
+            .emit_digest => {
+                const body_size = @sizeOf(std.zig.Server.Message.EmitDigest);
                 if (header.bytes_len <= body_size) return error.InvalidResponse;

-                _ = try zcs.receiveEmitBinPath();
+                _ = try zcs.receiveEmitDigest();

                 const trailing_size = header.bytes_len - body_size;
                 const result_path = zcs.pooler.fifo(.in).readableSliceOfLen(trailing_size);
```

**Comment by lmarzocchetti** [comment]
i have printed the communicated hash, but i don't find the algorithm used to decrypt this

**Comment by lmarzocchetti** [comment]
Okay i think this is it, the test is no longer failing, tell me if i'm wrong with the commit (i'm new to contributing to open source software)

File: src/translate_c.zig [hunk 2]
```diff
@@ -223,14 +223,16 @@ pub fn translate(
...
                 if (header.bytes_len <= body_size) return error.InvalidResponse;
 
-                _ = try zcs.receiveEmitBinPath();
+                _ = try zcs.receiveEmitDigest();
 
                 const trailing_size = header.bytes_len - body_size;
-                const result_path = zcs.pooler.fifo(.in).readableSliceOfLen(trailing_size);
+                const bin_result_path = zcs.pooler.fifo(.in).readableSliceOfLen(trailing_size);
+                const hex_result_path = std.Build.Cache.binToHex(@constCast(bin_result_path[0..16]).*);
+                const result_path = "/o/" ++ hex_result_path;
```
Comment (Techatrix) [inline src/translate_c.zig:235]: The test doesn't actually check whether the file actually exists or not. Obviously there is no file at `/o/` in your file system.
This patch should make the test fail if the file doesn't exist:

```patch
diff --git a/tests/language_features/cimport.zig b/tests/language_features/cimport.zig
index 44edf28d..018b5d68 100644
--- a/tests/language_features/cimport.zig
+++ b/tests/language_features/cimport.zig
@@ -117,6 +117,7 @@ fn testTranslate(c_source: []const u8) !translate_c.Result {
             const path = try zls.URI.parse(allocator, uri);
             defer allocator.free(path);
             try std.testing.expect(std.fs.path.isAbsolute(path));
+            try std.fs.accessAbsolute(path, .{});
         },
         .failure => |message| {
             try std.testing.expect(message.errorMessageCount() != 0);
```

I file should be located inside the `.zig-cache` directory like this: 
```zig
const result_path = try std.fs.path.join(allocator, &.{ global_cache_path, "o", &hex_result_path, "cimport.zig" });
defer allocator.free(result_path);
```

The only missing part is that we need to pass the `--cache-dir` argument when invoking zig:
```patch
diff --git a/src/translate_c.zig b/src/translate_c.zig
index ff7d383a..b7647650 100644
--- a/src/translate_c.zig
+++ b/src/translate_c.zig
@@ -156,6 +156,8 @@ pub fn translate(
         "translate-c",
         "--zig-lib-dir",
         zig_lib_path,
+        "--cache-dir",
+        global_cache_path,
         "--global-cache-dir",
         global_cache_path,
         "-lc",
```

Also, the `@constCast` is not necessary.

Comment (Techatrix) [inline src/translate_c.zig:235]: The test doesn't actually check whether the file actually exists or not. Obviously there is no file at `/o/` in your file system.
This patch should make the test fail if the file doesn't exist:

```patch
diff --git a/tests/language_features/cimport.zig b/tests/language_features/cimport.zig
index 44edf28d..018b5d68 100644
--- a/tests/language_features/cimport.zig
+++ b/tests/language_features/cimport.zig
@@ -117,6 +117,7 @@ fn testTranslate(c_source: []const u8) !translate_c.Result {
             const path = try zls.URI.parse(allocator, uri);
             defer allocator.free(path);
             try std.testing.expect(std.fs.path.isAbsolute(path));
+            try std.fs.accessAbsolute(path, .{});
         },
         .failure => |message| {
             try std.testing.expect(message.errorMessageCount() != 0);
```

I file should be located inside the `.zig-cache` directory like this: 
```zig
const result_path = try std.fs.path.join(allocator, &.{ global_cache_path, "o", &hex_result_path, "cimport.zig" });
defer allocator.free(result_path);
```

The only missing part is that we need to pass the `--cache-dir` argument when invoking zig:
```patch
diff --git a/src/translate_c.zig b/src/translate_c.zig
index ff7d383a..b7647650 100644
--- a/src/translate_c.zig
+++ b/src/translate_c.zig
@@ -156,6 +156,8 @@ pub fn translate(
         "translate-c",
         "--zig-lib-dir",
         zig_lib_path,
+        "--cache-dir",
+        global_cache_path,
         "--global-cache-dir",
         global_cache_path,
         "-lc",
```

Also, the `@constCast` is not necessary.

**Review by Techatrix — CHANGES_REQUESTED** [review]

File: src/translate_c.zig (hunk 2 above)
Comment (lmarzocchetti) [reply src/translate_c.zig]: Okay i finished, i have tested with a C library (raylib) and autocomplete is now working correctly!

**Comment by lmarzocchetti** [comment]
With the build "dffc8c4" of zig, zls can be builded seamlessy. I'have watched and there is a new release "7bbbbf8ff" and it fails to build entirely. I think is not my commit but something with zig itself

**Comment by Techatrix** [comment]
> With the build "dffc8c4" of zig, zls can be builded seamlessy. I'have watched and there is a new release "7bbbbf8ff" and it fails to build entirely. I think is not my commit but something with zig itself

That's https://github.com/ziglang/zig/pull/21135. It has fixed a bug with `@setEvalBranchQuota` that is also a breaking change.

**Comment by lmarzocchetti** [comment]
I have managed to build ZLS inserting @setEvalBranchQuota(3000) in the insertionContext function inside the standard library file "sort.zig". Idk if this is an useful information, but i wanted to let you know

**Comment by Techatrix** [comment]
The call to `@setEvalBranchQuota` should be where you reach into the standard library. One callsite is in the zig-lsp-codegen dependency of ZLS which is tracked [here](https://github.com/zigtools/zig-lsp-codegen/issues/16).

**Comment by lmarzocchetti** [comment]
Okay so we need to wait the change to zig-lsp-codegen. There is something that i can do, or just wait for the change?

**Comment by Techatrix** [comment]
The branch eval quota stuff should be taken care of. I have directly pushed to changes to this Pull Request.

@lmarzocchetti  FYI It is recommended to create a separate branch in your Fork (https://github.com/lmarzocchetti/zls) instead of committing to master when you plan to send a Pull Request.

**Review by Techatrix — APPROVED** [review]
Everything looks to be in order now.
Thank you @lmarzocchetti

**Comment by lmarzocchetti** [comment]
Thank you for your time helping me! And thank you for suggesting the branch stuff, next time i will be more prepared!

**Pull request merged**